### ISSUE TITLE
[Question] Perform Training based on a Subset of a Dataset

### ISSUE THREAD

**Issue by adebiasio-estilos** [RUCAIBox/RecBole#1582]
Hi All, first of all thank you very much for this great library!

I would like to perform model training based on a subset of a previously defined training set.

Currently I have tried changing the **inter_feat** variable in the dataset in the following way.

```
subset_percentage = 0.2

df_inter= pd.DataFrame({'user_id': dataset.inter_feat['user_id'], 
                                       'item_id': dataset.inter_feat['item_id'], 
                                       'label': dataset.inter_feat['label'])
df_inter_pos = self.df_inter_mask_item_feat.query("label == 1")

df_sample = df_inter_pos.sample(n=int(len(df_inter_pos )*subset_percentage))

df_inter.iloc[df_inter_pos.index, df_inter.columns.get_loc('label')] = 0

df_inter.iloc[df_sample.index, df_inter.columns.get_loc('label')] = 1

dataset.inter_feat = Interaction({'user_id': torch.tensor(df_inter['user_id'].values),
                                                  'item_id': torch.tensor(df_inter['item_id'].values),
                                                  'label': torch.tensor(df_inter['label'].values)})
```

However, the results of the model trained on this dataset are the same before and after defining the subset in this way.

Therefore I don't know if that is the correct way of defining a subset of a training set or not.

**Comment by Sherry-XLL** [comment]
Hello @adebiasio-estilos, thanks for your attention to our library!

Where do you modify the dataset code? Maybe the dataloader in `train_data` still uses the dataset before modification, resulting in no change in training results. If you only want to modify the training data, please modify the `dataset.inter_feat` fed to `train_data`, and then perform data iteration and training.

Feel free to contact us if you have any questions.

**Comment by adebiasio-estilos** [comment]
Hi @Sherry-XLL thank you very much for your help!

The code is use is the following:

```
def sampling(df_inter, batch_size, strategy="random", seed=123):
    df_sample = pd.DataFrame({})

    if strategy == 'random':
        df_sample = df_inter.sample(n=batch_size, random_state=seed)
    return df_sample

strategy="random"
mni = 0.2

config = Config(model="FM", dataset="yelp_recsys", config_dict=config_settings)
init_seed(config['seed'], config['reproducibility'])

init_logger(config)
logger = logging.getLogger()

dataset = create_dataset(config)
built_datasets = dataset.build()
train_dataset, valid_dataset, test_dataset = built_datasets
```

```
data = train_dataset

df_item_feat = pd.DataFrame({'item_id': data.item_feat['item_id'], 
                                  'category': data.item_feat['category'], 
                                  'subcategory': data.item_feat['subcategory'], 
                                  'price_bucket': data.item_feat['price_bucket']})

df_inter_mask = pd.DataFrame({'user_id': data.inter_feat['user_id'], 
                                   'item_id': data.inter_feat['item_id'], 
                                   'label': data.inter_feat['label'],
                                   'mask': 0})

df_inter_mask_item_feat = df_inter_mask.merge(df_item_feat, on='item_id')

df_inter_remaining_pos = df_inter_mask_item_feat.query("label == 1")

df_sample_pos = sampling(df_inter_remaining_pos, int(len(df_inter_remaining_pos)*mni), strategy=strategy, seed=config['seed'])

df_inter_mask_item_feat.iloc[df_inter_remaining_pos.index, df_inter_mask_item_feat.columns.get_loc('label')] = 0

df_inter_mask_item_feat.iloc[df_sample_pos.index, df_inter_mask_item_feat.columns.get_loc('label')] = 1

train_dataset.inter_feat = Interaction({'user_id': torch.tensor(df_inter_mask_item_feat['user_id'].values),
                        'item_id': torch.tensor(df_inter_mask_item_feat['item_id'].values),
                        'label': torch.tensor(df_inter_mask_item_feat['label'].values)})
```
```
built_datasets = [train_dataset, valid_dataset, test_dataset]

train_sampler, valid_sampler, test_sampler = create_samplers(
    config, dataset, built_datasets
)

train_data = get_dataloader(config, "train")(
    config, train_dataset, train_sampler, shuffle=config["shuffle"]
)
valid_data = get_dataloader(config, "evaluation")(
    config, valid_dataset, valid_sampler, shuffle=False
)
test_data = get_dataloader(config, "evaluation")(
    config, test_dataset, test_sampler, shuffle=False
)
```
```
model = FM(config, train_data.dataset).to(config['device'])
trainer = Trainer(config, model)
best_valid_score, best_valid_result = trainer.fit(train_data, valid_data, verbose=True, show_progress=False, saved=True)
test_result = trainer.evaluate(test_data, load_best_model=True, show_progress=False)
```
Do you see something `strange?

**Comment by Sherry-XLL** [comment]
Hello, @adebiasio-estilos! 

I modified the function `data_preparation` according to the code you provided. The running result is worse than before, indicating that the training dataset has indeed changed. There is nothing wrong with your modification, and you can check again whether the code that modifies the dataset is run as expected.

Thanks for your attention to RecBole.

**Comment by adebiasio-estilos** [comment]
Hi @Sherry-XLL, thanks for answering!

If you modify the `mni` parameter does the `_next_batch_data()` of the DataLoader change or not?

Because I found that if I change
`df_inter_mask_item_feat = df_inter_mask.merge(df_item_feat, on='item_id')
`
with
`df_inter_mask_item_feat = df_inter_mask.merge(df_item_feat, how="left", on='item_id')
`
and I explore the `_next_batch_data()` with
`pd.DataFrame(train_data._next_batch_data().numpy()).sort_values("user_id").query("label == 1")
`
the output does not change by varying `mni ` and I don't understand why..

**Comment by Sherry-XLL** [comment]
@adebiasio-estilos Hello, I changed the value of parameter `mni` and the results changed, so I don't know what your problem is. If you want to select a subset of the training set, removing the interactions may be better in general.

**Comment by adebiasio-estilos** [comment]
@Sherry-XLL thank you very much for answering!

Can you please share the code you used so I can check what I'm doing wrong?

**Comment by Sherry-XLL** [comment]
Hello @adebiasio-estilos, I just copied your code, and then experimented on dataset `ml-100k`, without making any changes to the code. If you get different results before and after using the function `sampling`, you should also get new results after changing the `mni` parameter.

**Comment by adebiasio-estilos** [comment]
Hi @Sherry-XLL, ok! Actually I get the same results before and after using the function sampling..

What is the version of RecBole you are using?

**Comment by adebiasio-estilos** [comment]
Hi @Sherry-XLL , you are right!
I've tried with that configuration and it works..
I should check deeper in my previous notebook what I was doing wrong

**Comment by adebiasio-estilos** [comment]
Hi @Sherry-XLL I am writing because I was finally able to investigate the problem more thoroughly!

I've prepared a Jupyter Notebook to illustrate the strange behavior I'm dealing with.

There are 3 experiments in the notebook:
- in the first, the classic `data_preparation` function found in RecBole is used
  - `AUC=0.5577 `
- in the second the `data_preparation_sampling` function (which was the one you had posted previously) with `mni=0.2` is used and the result is identical to the previous experiment
  - `AUC=0.5577 `
- in the third experiment, on the other hand, `data_preparation_sampling` with `mni=0.2` is used but no configuration is passed to the `Config` class (i.e., `Config(model="FM", dataset="ml-100k", config_file_list=None, config_dict=None)`) and the result is different from the previous two experiments (as it should be expected)
  - `AUC=0.496 `

I was careful to set almost all the parameters of the first two experiments to the default parameters of the third experiment.
The only notable difference in the configuration is the following:
- in the third experiment:
  - `"train_neg_sample_args" : {'distribution': 'none', 'sample_num': 'none', 'alpha': 'none', 'dynamic': False, 'candidate_num': 0}`
- in the first and second experiment
  - `"train_neg_sample_args" : {'distribution': 'uniform', 'sample_num': 1, 'dynamic': False}`

Do you know why is this happening?


[03 - Test RecBole.zip](https://github.com/RUCAIBox/RecBole/files/10315997/03.-.Test.RecBole.zip)
